src/rust/engine/fs/src/store.rs

```diff
@@ -0,0 +1,238 @@
+      |e| {
+        format!(
+          "Error creating/opening directories database: {}",
+          e.description()
+        )
+      },
+    )?;
+    Ok(Store {
+      env: env,
+      file_store: file_db_handle,
+      directory_store: directory_db_handle,
+    })
+  }
+
+  pub fn store_file_bytes(&self, bytes: &[u8]) -> Result<Fingerprint, String> {
+    self.store_bytes(bytes, &self.file_store)
+  }
+
+  fn store_bytes(&self, bytes: &[u8], db: &DbHandle) -> Result<Fingerprint, String> {
+    let mut hasher = Sha256::default();
+    hasher.input(&bytes);
+    let fingerprint = Fingerprint::from_bytes_unsafe(hasher.fixed_result().as_slice());
+
+    let txn = self.env.new_transaction().map_err(|e| {
+      format!(
+        "Error making new transaction to store fingerprint {}: {}",
+        fingerprint,
+        e.description()
+      )
+    })?;
+    {
+      let db = txn.bind(db);
+      match db.insert(&fingerprint, &bytes) {
+        Ok(_) => {}
+        Err(MdbError::KeyExists) => return Ok(fingerprint),
+        Err(err) => {
+          return Err(format!(
+            "Error storing fingerprint {}: {}",
+            fingerprint,
+            err.description()
+          ))
+        }
+      };
+    }
+    match txn.commit() {
+      Ok(_) => Ok(fingerprint),
+      Err(MdbError::KeyExists) => Ok(fingerprint),
+      Err(err) => Err(format!(
+        "Error committing transaction for fingerprint {}: {}",
+        fingerprint,
+        err.description()
+      )),
+    }
+  }
+
+  pub fn load_bytes(&self, fingerprint: &Fingerprint) -> Result<Option<Vec<u8>>, String> {
+    let reader = self.env.get_reader().map_err(|e| {
+      format!(
+        "Error getting reader for fingerprint {}: {}",
+        fingerprint,
+        e.description()
+      )
+    })?;
+    let db = reader.bind(&self.file_store);
+    match db.get(&fingerprint.as_bytes().as_ref()) {
+      Ok(v) => Ok(Some(v)),
+      Err(MdbError::NotFound) => Ok(None),
+      Err(err) => Err(format!(
+        "Error getting fingerprint {}: {}",
+        fingerprint,
+        err.description().to_string()
+      )),
+    }
+  }
+
+  ///
+  /// Store the Directory proto. Does not do anything about the files or directories claimed to be
+  /// contained therein.
+  ///
+  /// Assumes that the directory has been properly canonicalized.
+  ///
+  pub fn record_directory(
+    &self,
+    directory: &bazel_protos::remote_execution::Directory,
+  ) -> Result<(Fingerprint, usize), String> {
+    let bytes = directory.write_to_bytes().map_err(|e| {
+      format!(
+        "Error serializing directory proto {:?}: {}",
+        directory,
+        e.description()
+      )
+    })?;
+
+    Ok((
+      self.store_bytes(&bytes, &self.directory_store)?,
+      bytes.len(),
+    ))
+  }
+}
+
+impl ToMdbValue for Fingerprint {
+  fn to_mdb_value<'a>(&'a self) -> MdbValue<'a> {
+    unsafe { MdbValue::new(transmute(self.as_bytes().as_ptr()), self.as_bytes().len()) }
```

Comment:
It seems like there is an implementation of ToMdbValue for `&'a [u8]`: https://vhbit.github.io/lmdb-rs/lmdb_rs/traits/trait.ToMdbValue.html ... it doesn't look like this needs to be unsafe?

Reply:
That's an implementation for a slice, not an array. As far as I can tell, there's no way to turn an array into a slice which would have a long enough lifetime to be able to use the slice implementation. Very happy to be wrong about this, though!

Reply:
Hm. You might be right, but my understanding is certainly that you can slice any array (or Vector) with the same lifetime, since the slice is just a view into the underlying collection.

Reply:
From https://rustbyexample.com/primitives/array.html :
```
// Slices can point to a section of an array
println!("borrow a section of the array as a slice");
analyze_slice(&ys[1 .. 4]);
```

Reply:
The trait is implemented:
```
impl<'a> ToMdbValue for &'a [u8] {
    fn to_mdb_value<'b>(&'b self) -> MdbValue<'b> {
```

If I pass in a &[u8] I get told:
```
error[E0597]: borrowed value does not live long enough
   --> src/store.rs:119:42
    |
119 |     <&[u8] as ToMdbValue>::to_mdb_value(&&self.as_bytes()[..])
    |                                          ^^^^^^^^^^^^^^^^^^^^ does not live long enough
120 |   }
    |   - temporary value only lives until here
    |
note: borrowed value must be valid for the lifetime 'a as defined on the method body at 118:3...
   --> src/store.rs:118:3
    |
118 | /   fn to_mdb_value<'a>(&'a self) -> MdbValue<'a> {
119 | |     <&[u8] as ToMdbValue>::to_mdb_value(&&self.as_bytes()[..])
120 | |   }
    | |___^
```

If I pass in a [u8] I get told:
```
error[E0277]: the trait bound `[u8]: lmdb_rs::ToMdbValue` is not satisfied
   --> src/store.rs:119:5
    |
119 |     <[u8] as ToMdbValue>::to_mdb_value(&self.as_bytes()[..])
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `lmdb_rs::ToMdbValue` is not implemented for `[u8]`
    |
    = help: the following implementations were found:
              <&'a [u8] as lmdb_rs::ToMdbValue>
    = note: required by `lmdb_rs::ToMdbValue::to_mdb_value`
```

I tried attaching a lifetime to Fingerprint so I could return a slice of bytes bound to the lifetime of Fingerprint, but the moment you attach a lifetime to something, you need to change so much code to reflect that that it didn't seem worthwhile.

Got any pointers? :)

Reply:
...oh, weird. The trait does not really make sense then, afaict? Rather than:
```
impl<'a> ToMdbValue for &'a [u8] {
    fn to_mdb_value<'b>(&'b self) -> MdbValue<'b> {
```
it would need to be:
```
impl<'a> ToMdbValue for &'a [u8] {
    fn to_mdb_value(self) -> MdbValue<'a> {
```

As it stands, it requires a "reference to a reference", and there is nothing keeping the inner reference alive.

Reply:
So, it looks like there is another LDMB crate, with a potentially better API? https://docs.rs/lmdb/0.7.2/lmdb/struct.RwTransaction.html ... it uses `AsRef<[u8]>`, rather than its own odd `ToMdbValue` trait.

According to this blurb on the github repo for `lmdb-rs` (the crate used here), `lmdb` (the one I linked to above) has a better design: https://github.com/vhbit/lmdb-rs/issues/32#issuecomment-310906601

Reply:
Ugh. I didn't do a particularly thorough analysis of the lmdb crates available, I mostly saw that one had an order of magnitude more users than the others and grabbed it. I'll do a little more research...